[Quest API] Add Scripting Support to Mercenaries

Kinglykrab opened the pull request:
# Description
- Adds scripting support for Mercenaries using `merc.pl` or `global_merc.pl`.

# Events
- Mercenaries can use the following events.
```
EVENT_AGGRO_SAY
EVENT_CAST
EVENT_CAST_BEGIN
EVENT_CAST_ON
EVENT_COMBAT
EVENT_DAMAGE_GIVEN
EVENT_DAMAGE_TAKEN
EVENT_DEATH
EVENT_DEATH_COMPLETE
EVENT_DESPAWN
EVENT_ENTITY_VARIABLE_DELETE
EVENT_ENTITY_VARIABLE_SET
EVENT_ENTITY_VARIABLE_UPDATE
EVENT_NPC_SLAY
EVENT_PAYLOAD
EVENT_POPUPRESPONSE
EVENT_SAY
EVENT_SIGNAL
EVENT_SLAY
EVENT_SPAWN
EVENT_SPELL_BLOCKED
EVENT_TARGET_CHANGE
EVENT_TIMER
EVENT_TIMER_PAUSE
EVENT_TIMER_RESUME
EVENT_TIMER_START
EVENT_TIMER_STOP
EVENT_USE_SKILL
```

# Perl
- Add `$merc->GetCostFormula()`.
- Add `$merc->GetGroup()`.
- Add `$merc->GetHatedCount()`.
- Add `$merc->GetMaxMeleeRangeToTarget(target)`.
- Add `$merc->GetMercenaryCharacterID()`.
- Add `$merc->GetMercenaryID()`.
- Add `$merc->GetMercenaryNameType()`.
- Add `$merc->GetMercenaryOwner()`.
- Add `$merc->GetMercenarySubtype()`.
- Add `$merc->GetMercenaryTemplateID()`.
- Add `$merc->GetMercenaryType()`.
- Add `$merc->GetOwner()`.
- Add `$merc->GetOwnerOrSelf()`.
- Add `$merc->GetProficiencyID()`.
- Add `$merc->GetStance()`.
- Add `$merc->GetTierID()`.
- Add `$merc->HasOrMayGetAggro()`.
- Add `$merc->IsMercenaryCaster()`.
- Add `$merc->IsMercenaryCasterCombatRange(target)`.
- Add `$merc->IsSitting()`.
- Add `$merc->IsStanding()`.
- Add `$merc->ScaleStats(scale_percentage)`.
- Add `$merc->ScaleStats(scale_percentage, set_to_max)`.
- Add `$merc->SendPayload(payload_id, payload_value)`.
- Add `$merc->SetTarget(target)`.
- Add `$merc->Signal(signal_id)`.
- Add `$merc->Sit()`.
- Add `$merc->Stand()`.
- Add `$merc->Suspend()`.
- Add `$merc->UseDiscipline(spell_id, target_id)`.

## Perl Script
```pl
sub EVENT_CAST {
	quest::debug("spell_id " . $spell_id);
	quest::debug("caster_id " . $caster_id);
	quest::debug("caster_level " . $caster_level);
	quest::debug("target_id " . $target_id);
	quest::debug("target " . $target);
	quest::debug("spell " . $spell);
}

sub EVENT_CAST_BEGIN {
	quest::debug("spell_id " . $spell_id);
	quest::debug("caster_id " . $caster_id);
	quest::debug("caster_level " . $caster_level);
	quest::debug("target_id " . $target_id);
	quest::debug("target " . $target);
	quest::debug("spell " . $spell);
}

sub EVENT_CAST_ON {
	quest::debug("spell_id " . $spell_id);
	quest::debug("caster_id " . $caster_id);
	quest::debug("caster_level " . $caster_level);
	quest::debug("target_id " . $target_id);
	quest::debug("target " . $target);
	quest::debug("spell " . $spell);
}

sub EVENT_COMBAT {
	quest::debug("combat_state " . $combat_state);
}

sub EVENT_DAMAGE_GIVEN {
	quest::debug("entity_id " . $entity_id);
	quest::debug("damage " . $damage);
	quest::debug("spell_id " . $spell_id);
	quest::debug("skill_id " . $skill_id);
	quest::debug("is_damage_shield " . $is_damage_shield);
	quest::debug("is_avoidable " . $is_avoidable);
	quest::debug("buff_slot " . $buff_slot);
	quest::debug("is_buff_tic " . $is_buff_tic);
	quest::debug("special_attack " . $special_attack);
	quest::debug("spell " . $spell);
}

sub EVENT_DAMAGE_TAKEN {
	quest::debug("entity_id " . $entity_id);
	quest::debug("damage " . $damage);
	quest::debug("spell_id " . $spell_id);
	quest::debug("skill_id " . $skill_id);
	quest::debug("is_damage_shield " . $is_damage_shield);
	quest::debug("is_avoidable " . $is_avoidable);
	quest::debug("buff_slot " . $buff_slot);
	quest::debug("is_buff_tic " . $is_buff_tic);
	quest::debug("special_attack " . $special_attack);
	quest::debug("spell " . $spell);
}
sub EVENT_DEATH {
	quest::debug("killer_id " . $killer_id);
	quest::debug("killer_damage " . $killer_damage);
	quest::debug("killer_spell " . $killer_spell);
	quest::debug("killer_skill " . $killer_skill);
	quest::debug("killed_entity_id " . $killed_entity_id);
	quest::debug("combat_start_time " . $combat_start_time);
	quest::debug("combat_end_time " . $combat_end_time);
	quest::debug("damage_received " . $damage_received);
	quest::debug("healing_received " . $healing_received);
	quest::debug("killed_corpse_id " . $killed_corpse_id);
	quest::debug("killed_x " . $killed_x);
	quest::debug("killed_y " . $killed_y);
	quest::debug("killed_z " . $killed_z);
	quest::debug("killed_h " . $killed_h);
	quest::debug("killed_merc_id " . $killed_merc_id);
	quest::debug("killed_npc_id " . $killed_npc_id);
}

sub EVENT_DEATH_COMPLETE {
	quest::debug("killer_id " . $killer_id);
	quest::debug("killer_damage " . $killer_damage);
	quest::debug("killer_spell " . $killer_spell);
	quest::debug("killer_skill " . $killer_skill);
	quest::debug("killed_entity_id " . $killed_entity_id);
	quest::debug("combat_start_time " . $combat_start_time);
	quest::debug("combat_end_time " . $combat_end_time);
	quest::debug("damage_received " . $damage_received);
	quest::debug("healing_received " . $healing_received);
	quest::debug("killed_corpse_id " . $killed_corpse_id);
	quest::debug("killed_x " . $killed_x);
	quest::debug("killed_y " . $killed_y);
	quest::debug("killed_z " . $killed_z);
	quest::debug("killed_h " . $killed_h);
	quest::debug("killed_merc_id " . $killed_merc_id);
	quest::debug("killed_npc_id " . $killed_npc_id);
}

sub EVENT_DESPAWN {
	quest::debug("despawned_entity_id " . $despawned_entity_id);
	quest::debug("despawned_bot_id " . $despawned_bot_id);
	quest::debug("despawned_merc_id " . $despawned_merc_id);
	quest::debug("despawned_npc_id " . $despawned_npc_id);
}

sub EVENT_ENTITY_VARIABLE_DELETE {
	quest::debug("variable_name " . $variable_name);
	quest::debug("variable_value " . $variable_value);
}

sub EVENT_ENTITY_VARIABLE_SET {
	quest::debug("variable_name " . $variable_name);
	quest::debug("variable_value " . $variable_value);
}

sub EVENT_ENTITY_VARIABLE_UPDATE {
	quest::debug("variable_name " . $variable_name);
	quest::debug("new_value " . $new_value);
	quest::debug("old_value " . $old_value);
}

sub EVENT_NPC_SLAY {
	quest::debug("killed " . $killed);
	quest::debug("killed_npc " . $killed_npc);
}

sub EVENT_PAYLOAD {
	quest::debug("payload_id " . $payload_id);
	quest::debug("payload_value " . $payload_value);
}

sub EVENT_POPUPRESPONSE {
	quest::debug("popupid " . $popupid);
}

sub EVENT_SAY {
	quest::debug("data " . $data);
	quest::debug("text " . $text);
	quest::debug("langid " . $langid);
}

sub EVENT_SIGNAL {
	quest::debug("signal " . $signal);
}

sub EVENT_SLAY {
	quest::shout("Perl | EVENT_SLAY");
}

sub EVENT_SPAWN {
	quest::shout("Perl | EVENT_SPAWN");
}

sub EVENT_SPELL_BLOCKED {
	quest::debug("blocking_spell_id " . $blocking_spell_id);
	quest::debug("cast_spell_id " . $cast_spell_id);
	quest::debug("blocking_spell " . $blocking_spell);
	quest::debug("cast_spell " . $cast_spell);
}

sub EVENT_TARGET_CHANGE {
	quest::debug("target " . $target);
}

sub EVENT_TIMER {
	quest::debug("timer " . $timer);
}

sub EVENT_TIMER_PAUSE {
	quest::debug("timer " . $timer);
	quest::debug("duration " . $duration);
}

sub EVENT_TIMER_RESUME {
	quest::debug("timer " . $timer);
	quest::debug("duration " . $duration);
}

sub EVENT_TIMER_START {
	quest::debug("timer " . $timer);
	quest::debug("duration " . $duration);
}

sub EVENT_TIMER_STOP {
	quest::debug("timer " . $timer);
}

sub EVENT_USE_SKILL {
	quest::debug("skill_id " . $skill_id);
	quest::debug("skill_level " . $skill_level);
}
```

# Lua
- Add `merc:GetCostFormula()`.
- Add `merc:GetGroup()`.
- Add `merc:GetHatedCount()`.
- Add `merc:GetMaxMeleeRangeToTarget(target)`.
- Add `merc:GetMercenaryCharacterID()`.
- Add `merc:GetMercenaryID()`.
- Add `merc:GetMercenaryNameType()`.
- Add `merc:GetMercenaryOwner()`.
- Add `merc:GetMercenarySubtype()`.
- Add `merc:GetMercenaryTemplateID()`.
- Add `merc:GetMercenaryType()`.
- Add `merc:GetOwner()`.
- Add `merc:GetOwnerOrSelf()`.
- Add `merc:GetProficiencyID()`.
- Add `merc:GetStance()`.
- Add `merc:GetTierID()`.
- Add `merc:HasOrMayGetAggro()`.
- Add `merc:IsMercenaryCaster()`.
- Add `merc:IsMercenaryCasterCombatRange(target)`.
- Add `merc:IsSitting()`.
- Add `merc:IsStanding()`.
- Add `merc:ScaleStats(scale_percentage)`.
- Add `merc:ScaleStats(scale_percentage, set_to_max)`.
- Add `merc:SendPayload(payload_id, payload_value)`.
- Add `merc:SetTarget(target)`.
- Add `merc:Signal(signal_id)`.
- Add `merc:Sit()`.
- Add `merc:Stand()`.
- Add `merc:Suspend()`.
- Add `merc:UseDiscipline(spell_id, target_id)`.

## Lua Example
```lua
function event_cast(e)
	eq.debug("spell_id " .. tostring(e.spell_id));
	eq.debug("caster_id " .. tostring(e.caster_id));
	eq.debug("caster_level " .. tostring(e.caster_level));
	eq.debug("target_id " .. tostring(e.target_id));
	eq.debug("target " .. tostring(e.target));
	eq.debug("spell " .. tostring(e.spell));
end

function event_cast_begin(e)
	eq.debug("spell_id " .. tostring(e.spell_id));
	eq.debug("caster_id " .. tostring(e.caster_id));
	eq.debug("caster_level " .. tostring(e.caster_level));
	eq.debug("target_id " .. tostring(e.target_id));
	eq.debug("target " .. tostring(e.target));
	eq.debug("spell " .. tostring(e.spell));
end

function event_cast_on(e)
	eq.debug("spell_id " .. tostring(e.spell_id));
	eq.debug("caster_id " .. tostring(e.caster_id));
	eq.debug("caster_level " .. tostring(e.caster_level));
	eq.debug("target_id " .. tostring(e.target_id));
	eq.debug("target " .. tostring(e.target));
	eq.debug("spell " .. tostring(e.spell));
end

function event_combat(e)
	eq.debug("combat_state " .. tostring(e.combat_state));
end

function event_damage_given(e)
	eq.debug("entity_id " .. tostring(e.entity_id));
	eq.debug("damage " .. tostring(e.damage));
	eq.debug("spell_id " .. tostring(e.spell_id));
	eq.debug("skill_id " .. tostring(e.skill_id));
	eq.debug("is_damage_shield " .. tostring(e.is_damage_shield));
	eq.debug("is_avoidable " .. tostring(e.is_avoidable));
	eq.debug("buff_slot " .. tostring(e.buff_slot));
	eq.debug("is_buff_tic " .. tostring(e.is_buff_tic));
	eq.debug("special_attack " .. tostring(e.special_attack));
	eq.debug("spell " .. tostring(e.spell));
end

function event_damage_taken(e)
	eq.debug("entity_id " .. tostring(e.entity_id));
	eq.debug("damage " .. tostring(e.damage));
	eq.debug("spell_id " .. tostring(e.spell_id));
	eq.debug("skill_id " .. tostring(e.skill_id));
	eq.debug("is_damage_shield " .. tostring(e.is_damage_shield));
	eq.debug("is_avoidable " .. tostring(e.is_avoidable));
	eq.debug("buff_slot " .. tostring(e.buff_slot));
	eq.debug("is_buff_tic " .. tostring(e.is_buff_tic));
	eq.debug("special_attack " .. tostring(e.special_attack));
	eq.debug("spell " .. tostring(e.spell));
end

function event_death(e)
	eq.debug("killer_id " .. tostring(e.killer_id));
	eq.debug("killer_damage " .. tostring(e.killer_damage));
	eq.debug("killer_spell " .. tostring(e.killer_spell));
	eq.debug("killer_skill " .. tostring(e.killer_skill));
	eq.debug("killed_entity_id " .. tostring(e.killed_entity_id));
	eq.debug("combat_start_time " .. tostring(e.combat_start_time));
	eq.debug("combat_end_time " .. tostring(e.combat_end_time));
	eq.debug("damage_received " .. tostring(e.damage_received));
	eq.debug("healing_received " .. tostring(e.healing_received));
	eq.debug("killed_corpse_id " .. tostring(e.killed_corpse_id));
	eq.debug("killed_x " .. tostring(e.killed_x));
	eq.debug("killed_y " .. tostring(e.killed_y));
	eq.debug("killed_z " .. tostring(e.killed_z));
	eq.debug("killed_h " .. tostring(e.killed_h));
	eq.debug("killed_merc_id " .. tostring(e.killed_merc_id));
	eq.debug("killed_npc_id " .. tostring(e.killed_npc_id));
end

function event_death_complete(e)
	eq.debug("killer_id " .. tostring(e.killer_id));
	eq.debug("killer_damage " .. tostring(e.killer_damage));
	eq.debug("killer_spell " .. tostring(e.killer_spell));
	eq.debug("killer_skill " .. tostring(e.killer_skill));
	eq.debug("killed_entity_id " .. tostring(e.killed_entity_id));
	eq.debug("combat_start_time " .. tostring(e.combat_start_time));
	eq.debug("combat_end_time " .. tostring(e.combat_end_time));
	eq.debug("damage_received " .. tostring(e.damage_received));
	eq.debug("healing_received " .. tostring(e.healing_received));
	eq.debug("killed_corpse_id " .. tostring(e.killed_corpse_id));
	eq.debug("killed_x " .. tostring(e.killed_x));
	eq.debug("killed_y " .. tostring(e.killed_y));
	eq.debug("killed_z " .. tostring(e.killed_z));
	eq.debug("killed_h " .. tostring(e.killed_h));
	eq.debug("killed_merc_id " .. tostring(e.killed_merc_id));
	eq.debug("killed_npc_id " .. tostring(e.killed_npc_id));
end

function event_despawn(e)
	eq.debug("despawned_entity_id " .. tostring(e.despawned_entity_id));
	eq.debug("despawned_bot_id " .. tostring(e.despawned_bot_id));
	eq.debug("despawned_merc_id " .. tostring(e.despawned_merc_id));
	eq.debug("despawned_npc_id " .. tostring(e.despawned_npc_id));
end

function event_entity_variable_delete(e)
	eq.debug("variable_name " .. tostring(e.variable_name));
	eq.debug("variable_value " .. tostring(e.variable_value));
end

function event_entity_variable_set(e)
	eq.debug("variable_name " .. tostring(e.variable_name));
	eq.debug("variable_value " .. tostring(e.variable_value));
end

function event_entity_variable_update(e)
	eq.debug("variable_name " .. tostring(e.variable_name));
	eq.debug("new_value " .. tostring(e.new_value));
	eq.debug("old_value " .. tostring(e.old_value));
end

function event_npc_slay(e)
	eq.debug("killed " .. tostring(e.killed));
	eq.debug("killed_npc " .. tostring(e.killed_npc));
end

function event_payload(e)
	eq.debug("payload_id " .. tostring(e.payload_id));
	eq.debug("payload_value " .. tostring(e.payload_value));
end

function event_popupresponse(e)
	eq.debug("popupid " .. tostring(e.popupid));
end

function event_say(e)
	eq.debug("data " .. tostring(e.data));
	eq.debug("text " .. tostring(e.text));
	eq.debug("langid " .. tostring(e.langid));
end

function event_signal(e)
	eq.debug("signal " .. tostring(e.signal));
end

function event_slay(e)
	eq.shout("Lua | EVENT_SLAY");
end

function event_spawn(e)
	eq.shout("Lua | EVENT_SPAWN");
end

function event_spell_blocked(e)
	eq.debug("blocking_spell_id " .. tostring(e.blocking_spell_id));
	eq.debug("cast_spell_id " .. tostring(e.cast_spell_id));
	eq.debug("blocking_spell " .. tostring(e.blocking_spell));
	eq.debug("cast_spell " .. tostring(e.cast_spell));
end

function event_target_change(e)
	eq.debug("target " .. tostring(e.target));
end

function event_timer(e)
	eq.debug("timer " .. tostring(e.timer));
end

function event_timer_pause(e)
	eq.debug("timer " .. tostring(e.timer));
	eq.debug("duration " .. tostring(e.duration));
end

function event_timer_resume(e)
	eq.debug("timer " .. tostring(e.timer));
	eq.debug("duration " .. tostring(e.duration));
end

function event_timer_start(e)
	eq.debug("timer " .. tostring(e.timer));
	eq.debug("duration " .. tostring(e.duration));
end

function event_timer_stop(e)
	eq.debug("timer " .. tostring(e.timer));
end

function event_use_skill(e)
	eq.debug("skill_id " .. tostring(e.skill_id));
	eq.debug("skill_level " .. tostring(e.skill_level));
end
```

# Testing
## EVENT_CAST | EVENT_CAST_BEGIN | EVENT_CAST_ON
![image](https://github.com/user-attachments/assets/424602fc-35ec-48f6-a1e6-5c251ff86cd5)
## EVENT_COMBAT
![image](https://github.com/user-attachments/assets/24cb1376-dab0-4c26-865e-05304defb8cd)
## EVENT_DAMAGE_GIVEN | EVENT_DAMAGE_TAKEN
![image](https://github.com/user-attachments/assets/ab8ce436-32f8-407c-9874-0c947265c064)
## EVENT_DEATH_COMPLETE
![image](https://github.com/user-attachments/assets/37b4ab96-33ce-4718-94d1-bbdf8d22b9b6)
### EVENT_DESPAWN
![image](https://github.com/user-attachments/assets/034fe2e3-cf84-4a8a-8b33-ee9ec3a4e99e)
## EVENT_ENTITY_VARIABLE_DELETE
![image](https://github.com/user-attachments/assets/e695086a-e498-45fe-bb56-13f42f2d4074)
## EVENT_ENTITY_VARIABLE_SET
![image](https://github.com/user-attachments/assets/b18d37c3-d359-4184-883a-abe69067fd15)
## EVENT_ENTITY_VARIABLE_UPDATE
![image](https://github.com/user-attachments/assets/2d9fa09f-475c-418f-99c8-fccb18eb85b9)
## EVENT_NPC_SLAY
![image](https://github.com/user-attachments/assets/23d69bc1-d324-48dd-afd8-d2e7be70fc09)
## EVENT_PAYLOAD
![image](https://github.com/user-attachments/assets/b048ab86-938f-45e1-ab9e-33d903f69a14)
## EVENT_POPUPRESPONSE
![image](https://github.com/user-attachments/assets/8d1fa137-b52b-439c-b31a-f4dd62e00f41)
## EVENT_SAY
![image](https://github.com/user-attachments/assets/e205a621-51ce-4535-bfe9-72f6120371ba)
## EVENT_SIGNAL
![image](https://github.com/user-attachments/assets/05d6a1b1-8a9b-440b-b749-c4bd806d30fb)
## EVENT_SPAWN
![image](https://github.com/user-attachments/assets/6e04918c-853e-4741-9569-fc398256e84b)
## EVENT_SPELL_BLOCKED
![image](https://github.com/user-attachments/assets/5b622199-92f2-4219-bb6f-ddc2f0373caf)
## EVENT_TIMER | EVENT_TIMER_STOP
![image](https://github.com/user-attachments/assets/140394f8-f820-4dd1-9d00-bec40e87ed76)
## EVENT_TIMER_PAUSE
![image](https://github.com/user-attachments/assets/ea93bce7-1091-4e1c-b1da-4fd7575415d8)
## EVENT_TIMER_RESUME
![image](https://github.com/user-attachments/assets/97e45fe0-986a-4e00-ada5-82b3a12ebfed)
## EVENT_TIMER_START
![image](https://github.com/user-attachments/assets/f7181b14-0b12-4836-9bd3-a8e7dbd651a2)
## EVENT_USE_SKILL
![image](https://github.com/user-attachments/assets/4ac30927-f2a2-4c52-85a2-4b928d74732d)

Akkadius commented on zone/embparser.cpp:
```diff
@@ -1101,19 +1126,21 @@ int PerlembParser::SendCommands(
-			clear_vars_[sp]  = 1;
-			clear_vars_[enl] = 1;
-			clear_vars_[bot] = 1;
+			std::string cl   = fmt::format("${}::client", prefix);
```
Can simplify (untested)

```cpp
std::vector<std::string> suffixes = {"client", "npc", "questitem", "spell", "entity_list", "bot", "merc"};

for (const auto& suffix : suffixes) {
    std::string key = fmt::format("${}::{}", prefix, suffix);
    clear_vars_[key] = 1;
}
```

Akkadius commented on zone/lua_merc.h:
```diff
@@ -0,0 +1,64 @@
+\
```
Backslash ?

Akkadius commented on zone/lua_merc.h:
```diff
@@ -0,0 +1,64 @@
+	bool UseDiscipline(uint16 spell_id, uint16 target_id);
+};
+
+#endif
+#endif
```
```cpp
#endif // LUA_EQEMU
#endif // EQEMU_LUA_MERC_H
```

Akkadius commented on zone/npc.cpp:
```diff
@@ -833,25 +833,33 @@ void NPC::Depop(bool start_spawn_timer) {
 	}
 
-	p_depop = true;
+		p_depop = true;
```
Accidental tab looks like

Akkadius commented:
Added code to simplify event invocation. Here is some of my attached testing

**Timers**

![image](https://github.com/user-attachments/assets/ef9db626-15ea-45fd-bfec-572d8248d16c)

```
function event_spawn(e)
	eq.set_timer("test", 1000);
end

function event_timer(e)
	eq.debug("timer " .. e.timer);
	e.self:Say("Test");
end
```

**Say**

![image](https://github.com/user-attachments/assets/c7edd087-ae5e-48e1-bdec-356267acfb67)

**Death**

![image](https://github.com/user-attachments/assets/ad13ad3f-f1cb-46f7-bd0b-6ee86415f948)

```lua
function event_death(e)
    e.self:Say("Oooof, I died! It sucks!")
    eq.debug("killer_id " .. e.killer_id);
    eq.debug("damage " .. e.damage);
    eq.debug("skill_id " .. e.skill_id);
    eq.debug("killed_entity_id " .. e.killed_entity_id);
    eq.debug("combat_start_time " .. e.combat_start_time);
    eq.debug("combat_end_time " .. e.combat_end_time);
    eq.debug("damage_received " .. e.damage_received);
    eq.debug("healing_received " .. e.healing_received);
end

Akkadius commented:
EVENT_TIMER_START

![image](https://github.com/user-attachments/assets/fb4d2270-631a-4bcd-b675-1f788e80ba64)

hgtw commented on zone/npc.cpp:
```diff
@@ -833,23 +833,7 @@ void NPC::Depop(bool start_spawn_timer) {
-			DispatchZoneControllerEvent(EVENT_DESPAWN_ZONE, this, "", 0, nullptr);
-		}
-	}
+	parse->EventNPCBotMerc(EVENT_DESPAWN, this);
```
Zone controller EVENT_DESPAWN_ZONE was removed here but not added back

hgtw commented on zone/quest_parser_collection.h:
```diff
@@ -126,6 +127,167 @@ class QuestParserCollection {
+
+		return false; // No quest subscription found
+	}
+
```
The definitions for all these new functions should be placed in the .cpp instead of this header

hgtw commented on zone/questmgr.cpp:
```diff
@@ -573,21 +549,17 @@ void QuestManager::settimer(const std::string& timer_name, uint32 seconds, Mob*
+				"{} {}",
+				timer_name,
+				seconds * 1000
+			);
```
It looks like this lambda is re-used without the arguments changing, so it could just be defined once before the first EVENT_TIMER_START and passed to both `parse->EventPlayerNPCBotMerc(EVENT_TIMER_START, mob, exportfn);`

Same with all the other timer events below

hgtw commented on zone/spells.cpp:
```diff
@@ -254,52 +254,33 @@ bool Mob::CastSpell(uint16 spell_id, uint16 target_id, CastingSlot slot,
+	);
+
+	if (IsClient()) {
+		if (return_value != 0) {
```
nit: combine the condition checks to reduce an indent level

hgtw commented on zone/spells.cpp:
```diff
@@ -3619,45 +3576,30 @@ int Mob::AddBuff(Mob *caster, uint16 spell_id, int duration, int32 level_overrid
+							);
 						}
-					}
+					);
```
This call is missing the `caster` argument for initiator that the old code had. It's also missing the  `if (caster != this)` check that prevented the caster from getting double events if casting on self

The duplicate lambda could also be re-used to turn both EVENT_SPELL_BLOCKED calls into one liners.

hgtw reviewed (commented):
Main notes:
EVENT_DAMAGE_TAKEN/EVENT_DAMAGE_GIVEN look broken
EVENT_SPELL_BLOCKED has a possible issues with 2nd call
EVENT_DESPAWN_ZONE for zone controller was removed
`EventPlayerNPCBotMerc` should probably just be named `EventMob` since it's so generic

Just brainstorming future changes (not this pr), think it would be better to ditch using these export strings (and maybe the extra_data int since it's rarely used) for events completely and just pass all event args in the `std::vector<std::any>`. Then quest parsers could handle args from a single source instead of dealing with janky and costly string construction/separator parsing. Especially for events with int args that end up converting from int to string then back to int for the export.

Kinglykrab commented:
> Main notes: EVENT_DAMAGE_TAKEN/EVENT_DAMAGE_GIVEN look broken EVENT_SPELL_BLOCKED has a possible issues with 2nd call EVENT_DESPAWN_ZONE for zone controller was removed `EventPlayerNPCBotMerc` should probably just be named `EventMob` since it's so generic
> 
> Just brainstorming future changes (not this pr), think it would be better to ditch using these export strings (and maybe the extra_data int since it's rarely used) for events completely and just pass all event args in the `std::vector<std::any>`. Then quest parsers could handle args from a single source instead of dealing with janky and costly string construction/separator parsing. Especially for events with int args that end up converting from int to string then back to int for the export.

- Fixed the damage events.
- Fixed `EVENT_SPELL_BLOCKED` calls.
- Added back `EVENT_DESPAWN_ZONE`.

I think removing them and just using the `std::vector<std::any>` definitely makes sense, that way it's all contained in one place and doesn't make it harder to understand what some of these events are doing in the future.

hgtw commented on zone/attack.cpp:
```diff
@@ -4285,100 +4257,60 @@ void Mob::CommonDamage(Mob* attacker, int64 &damage, const uint16 spell_id, cons
+			[&]() {
+				return fmt::format(
+					"{} {} {} {} {} {} {} {} {}",
+					attacker->GetID(),
```
This needs guarded against nullptr attacker like the old code `attacker ? attacker->GetID() : 0,`

hgtw commented on zone/embparser.cpp:
```diff
@@ -2060,14 +2085,16 @@ void PerlembParser::ExportEventVariables(
-					ExportVar(package_name.c_str(), "killed_y", corpse->GetY());
-					ExportVar(package_name.c_str(), "killed_z", corpse->GetZ());
-					ExportVar(package_name.c_str(), "killed_h", corpse->GetHeading());
+			if (!npc_mob->IsMerc()) {
```
This check shouldn't be necessary now since corpse will be nullptr for mercs and not do anything

hgtw commented on zone/lua_parser_events.cpp:
```diff
@@ -2789,4 +2801,380 @@ void handle_bot_spell_blocked(
+	lua_setfield(L, -2, "timer");
+}
+
+void handle_merc_use_skill(
```
All these unused handle_merc functions can be removed now

hgtw commented on zone/spells.cpp:
```diff
@@ -3619,45 +3574,17 @@ int Mob::AddBuff(Mob *caster, uint16 spell_id, int duration, int32 level_overrid
-							}
-						}
-					}
+					parse->EventMob(EVENT_SPELL_BLOCKED, this, caster, f);
```
This second event call still needs to only be done if `caster != this` like the old code to prevent double events (presumably if caster is casting on self)
